src/platform/windows/ipc/ipc_session.cpp

```diff
@@ -0,0 +1,267 @@
+
+/**
+ * @file ipc_session.cpp
+ * @brief Implements the IPC session logic for Windows WGC capture integration.
+ *
+ * Handles inter-process communication, shared texture setup, and frame synchronization
+ * between the main process and the WGC capture helper process.
+ */
+
+// standard includes
+#include <chrono>
+#include <filesystem>
+#include <string_view>
+#include <thread>
+
+// local includes
+#include "config.h"
+#include "ipc_session.h"
+#include "misc_utils.h"
+#include "src/logging.h"
+#include "src/platform/windows/misc.h"
+
+// platform includes
+#include <avrt.h>
+
+namespace platf::dxgi {
+
+  void ipc_session_t::handle_shared_handle_message(const std::vector<uint8_t> &msg, bool &handle_received) {
+    if (msg.size() == sizeof(shared_handle_data_t)) {
+      shared_handle_data_t handle_data;
+      memcpy(&handle_data, msg.data(), sizeof(shared_handle_data_t));
+      if (setup_shared_texture(handle_data.texture_handle, handle_data.width, handle_data.height)) {
+        handle_received = true;
+      }
+    }
+  }
+
+  void ipc_session_t::handle_frame_notification(const std::vector<uint8_t> &msg) {
+    if (msg.size() == sizeof(frame_ready_msg_t)) {
+      frame_ready_msg_t frame_msg;
+      memcpy(&frame_msg, msg.data(), sizeof(frame_msg));
+
+      if (frame_msg.message_type == FRAME_READY_MSG) {
+        _frame_qpc.store(frame_msg.frame_qpc, std::memory_order_release);
+        _frame_ready.store(true, std::memory_order_release);
+      }
+    }
+  }
+
+  void ipc_session_t::handle_secure_desktop_message(const std::vector<uint8_t> &msg) {
+    if (msg.size() == 1 && msg[0] == SECURE_DESKTOP_MSG) {
+      // secure desktop detected
+      BOOST_LOG(info) << "WGC can no longer capture the screen due to Secured Desktop, swapping to DXGI";
+      _should_swap_to_dxgi = true;
+    }
+  }
+
+
+  int ipc_session_t::init(const ::video::config_t &config, std::string_view display_name, ID3D11Device *device) {
+    _process_helper = std::make_unique<ProcessHandler>();
+    _config = config;
+    _display_name = display_name;
+    _device = device;
+    return 0;
+  }
+
+  void ipc_session_t::initialize_if_needed() {
+    if (_initialized.load()) {
+      return;
+    }
+
+    if (bool expected = false; !_initialized.compare_exchange_strong(expected, true)) {
+      // Another thread is already initializing, wait for it to complete
+      while (_initialized.load()) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      }
+      return;
+    }
+
+    // Check if properly initialized via init() first
+    if (!_process_helper) {
+      BOOST_LOG(debug) << "Cannot lazy_init without proper initialization";
+      return;
+    }
+
+    // Get the directory of the main executable
+    std::string exePathBuffer(MAX_PATH, '\0');
+    GetModuleFileNameA(nullptr, exePathBuffer.data(), MAX_PATH);
```

Comment:
We should be using the wide/unicode character variants of these, since we could be installed in a directory that contains non-ANSI characters.